Hoisting common expressions out of If blocks [retry]

Gamrix opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #65645

Summary:
This is a retry of PR: https://github.com/pytorch/pytorch/pull/59492

Latest Changes: Added more tests, added the getOrCreateDB pattern, updated logic to remove unnecessary checks
addressed all comments.

Adding code to find common expressions from the two subblocks of an if
operation and hoist them before the if block.
This also allows Dead Code Elimination to
then eliminate some if blocks.


Test Plan:
python test_jit.py TestIfHoisting

Differential Revision: [D33302065](https://our.internmc.facebook.com/intern/diff/D33302065)

Gamrix committed:
Hoisting common expressions out of If blocks [retry]

Summary:
Latest Changes: Added more tests, added the getOrCreateDB pattern, updated logic to remove unnecessary checks
addressed all comments.

Adding code to find common expressions from the two subblocks of an if
operation and hoist them before the if block.
This also allows Dead Code Elimination to
then eliminate some if blocks.

Also eliminated some dead code in the codebase.

Test Plan:
python test_jit.py TestIfHoisting

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Gamrix committed:
Update on "Hoisting common expressions out of If blocks [retry]"


Summary:
Latest Changes: Added more tests, added the getOrCreateDB pattern, updated logic to remove unnecessary checks
addressed all comments.

Adding code to find common expressions from the two subblocks of an if
operation and hoist them before the if block.
This also allows Dead Code Elimination to
then eliminate some if blocks.

Also eliminated some dead code in the codebase.

Test Plan:
python test_jit.py TestIfHoisting

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Gamrix commented:
@gamrix has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33302065).

eellison commented on torch/csrc/jit/passes/symbolic_shape_analysis.cpp:
```diff
@@ -6,6 +6,7 @@
 #include <torch/csrc/jit/ir/ir.h>
 #include <torch/csrc/jit/ir/ir_views.h>
 #include <torch/csrc/jit/jit_log.h>
+#include <torch/csrc/jit/passes/common_expression_hoisting.h>
```
nit, dont need this here

eellison reviewed (approved):
stamp

Gamrix committed:
Update on "Hoisting common expressions out of If blocks [retry]"


Summary:
This is a retry of PR: https://github.com/pytorch/pytorch/pull/59492

Latest Changes: Added more tests, added the getOrCreateDB pattern, updated logic to remove unnecessary checks
addressed all comments.

Adding code to find common expressions from the two subblocks of an if
operation and hoist them before the if block.
This also allows Dead Code Elimination to
then eliminate some if blocks.


Test Plan:
python test_jit.py TestIfHoisting

Differential Revision: [D33302065](https://our.internmc.facebook.com/intern/diff/D33302065)

[ghstack-poisoned]

Gamrix committed:
Update on "Hoisting common expressions out of If blocks [retry]"


Summary:
This is a retry of PR: https://github.com/pytorch/pytorch/pull/59492

Latest Changes: Added more tests, added the getOrCreateDB pattern, updated logic to remove unnecessary checks
addressed all comments.

Adding code to find common expressions from the two subblocks of an if
operation and hoist them before the if block.
This also allows Dead Code Elimination to
then eliminate some if blocks.


Test Plan:
python test_jit.py TestIfHoisting

Differential Revision: [D33302065](https://our.internmc.facebook.com/intern/diff/D33302065)

[ghstack-poisoned]

Gamrix commented:
@gamrix has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33302065).

Gamrix committed:
Update on "Hoisting common expressions out of If blocks [retry]"


Summary:
This is a retry of PR: https://github.com/pytorch/pytorch/pull/59492

Latest Changes: Added more tests, added the getOrCreateDB pattern, updated logic to remove unnecessary checks
addressed all comments.

Adding code to find common expressions from the two subblocks of an if
operation and hoist them before the if block.
This also allows Dead Code Elimination to
then eliminate some if blocks.


Test Plan:
python test_jit.py TestIfHoisting

Differential Revision: [D33302065](https://our.internmc.facebook.com/intern/diff/D33302065)

[ghstack-poisoned]
